### PR TITLE
Otp generation using redis

### PR DESCRIPTION
**Pull request by Ayushnit**

### CODE REVIEW

File: cmd/migrate/main.go [hunk 1]
```diff
@@ -13,6 +14,7 @@ import (
 var (
 	flags = flag.NewFlagSet("migrate", flag.ExitOnError)
 	dir   = flags.String("dir", "../../migration", "directory with migration files")
+	err_ error
```
Comment (mukesh0513) [inline cmd/migrate/main.go]: why is this needed as global variable?

File: cmd/migrate/main.go [hunk 2]
```diff
@@ -41,6 +43,11 @@ func main() {
 	if err := goose.Run(command, db, *dir); err != nil {
 		log.Fatalf("goose %v: %v", command, err)
 	}
+	config.DB,err_=gorm.Open("mysql",config.DbURL(config.BuildConfig()))
```
Comment (mukesh0513) [inline cmd/migrate/main.go]: not needed. db initialisation has already happened in config.go

File: internal/provider/redis/redis.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"crypto/rand"
+	"github.com/go-redis/redis"
+	"math/big"
+	"strconv"
+	"time"
+)
+
+var redisClient = redis.NewClient(&redis.Options{
```
Comment (mukesh0513) [inline internal/provider/redis/redis.go]: Add createClient functionality in the method itself.

File: internal/provider/redis/redis.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	Addr:     "localhost:6379",
```
Comment (mukesh0513) [inline internal/provider/redis/redis.go]: Add this data into configuration.

File: internal/provider/redis/redis.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+})
+
+func SetValue(key string, value string, expiry time.Duration) error {
+	err := redisClient.Set(key, value, expiry).Err()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func GetRandNum() (string, error) {
```
Comment (mukesh0513) [inline internal/provider/redis/redis.go]: does this method needs to be in redisProvider?

File: internal/provider/redis/redis.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	nBig, e := rand.Int(rand.Reader, big.NewInt(8999))
+	if e != nil {
+		return "", e
+	}
+	return strconv.FormatInt(nBig.Int64()+1000, 10), nil
+}
+
+func GetValue(key string) (string, error) {
+	value, argh := redisClient.Get(key).Result()
```
Comment (mukesh0513) [inline internal/provider/redis/redis.go]: what is argh?

File: internal/router/routes.go [hunk 7]
```diff
@@ -37,5 +37,14 @@ func SetUpRouter() *gin.Engine  {
 	//	websocket.GET("/stocks", webSocket.StockHandler)
 	//	websocket.GET("/orders", webSocket.OrderHandler)
 	//}
+
+	users := r.Group("/users")
+	{
+		users.POST("/signup", controller.SignUp)
+		users.POST("/:Userid/details", controller.UserDetails)
+		users.POST("/signIn", controller.SignIn)
```
Comment (mukesh0513) [inline internal/router/routes.go]: these are not variables. add snake_case.

File: internal/services/stock_exchange/stock_exchange.go [hunk 8]
```diff
@@ -1,5 +1,22 @@
 package stock_exchange
 
+import (
+	"github.com/jinzhu/gorm"
+	"log"
+)
+
+var DB *gorm.DB
```
Comment (mukesh0513) [inline internal/services/stock_exchange/stock_exchange.go]: Why is this object defined here? This already has been initialised in config file.

Comment (mukesh0513) [reply internal/services/stock_exchange/stock_exchange.go]: Why this file is present in your PR?

File: internal/services/user_management/user_management.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	"Tradeasy/internal/model"
+	"Tradeasy/internal/provider/redis"
+	"errors"
+	"strings"
+	"time"
+)
+
+func SignUp(SUpReq SignUpRequest) (SUpRes SignUpResponse, err error) {
+	email := SUpReq.EmailId
+	SUpRes.UserId = strings.Split(email, "@")[0]
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Why is this splitting needed?

File: internal/services/user_management/user_management.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+	SUpRes.Password = SUpReq.Password
+	SUpRes.Message = "User registered"
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Create response at the end of method if everything works fine.

File: internal/services/user_management/user_management.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+	var user model.Users
+	err = config.DB.Table("users").Where("email_id = ?", SUpReq.EmailId).First(&user).Error
+	if err == nil {
+		return SUpRes, errors.New("email id already registered")
+	}
+	err = config.DB.Table("users").Where("password = ?", SUpRes.Password).First(&user).Error
+	if err == nil {
+		return SUpRes, errors.New("password already taken")
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: You have already filled your response message with 'user registered'

File: internal/services/user_management/user_management.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+	user.Password = SUpReq.Password
+	err = config.DB.Table("users").Create(&user).Error
+	if err != nil {
+		return SUpRes, errors.New("signUp failed")
+	}
+	return SUpRes, nil
+}
+
+func UserDetails(detReq UserDetailsRequest, userid string) (detRes UserDetailsResponse, err error) {
+	detRes.TradingAccId = "TRA" + userid
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: what is TRA?

File: internal/services/user_management/user_management.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	detRes.Balance = 0
+	detRes.Message = "User Details registered"
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Do not add message before request gets successfully processed.

File: internal/services/user_management/user_management.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	var user model.Users
+	err = config.DB.Table("users").Where("user_id = ? ", userid).First(&user).Error
+	if err != nil {
+		return detRes, errors.New("user not found")
+	}
+	var ta model.TradingAccount
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Initialise model at the method start for better readability.

File: internal/services/user_management/user_management.go (hunk 12 above)
Comment (Ayushnit) [reply internal/services/user_management/user_management.go]: I have just appended TRA to the user id to make the trading account id

File: internal/services/user_management/user_management.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+	err = config.DB.Table("trading_account").Where("user_id = ? OR pan_card_no = ? OR bank_acc_no = ?", userid, detReq.PanCardNo, detReq.BankAccNo).First(&ta).Error
+	if err == nil {
+		return detRes, errors.New("user details already registered")
+	}
+	ta.UserId = userid
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Initialise as struct instead of each separate lines.

File: internal/services/user_management/user_management.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	ta.PanCardNo = detReq.PanCardNo
+	ta.BankAccNo = detReq.BankAccNo
+	ta.TradingAccId = detRes.TradingAccId
+	ta.Balance = detRes.Balance
+
+	err = config.DB.Table("trading_account").Create(&ta).Error
+	if err != nil {
+		return detRes, errors.New("enter correct user details")
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Message does not tells the exact error.

File: internal/services/user_management/user_management.go [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	return detRes, nil
+}
+func UserSignIn(SInReq SignInRequest) (SInRes SignInResponse, err error) {
+	SInRes.Message = "Signed in successfully"
+
+	var user model.Users
+	err = config.DB.Table("users").Where("user_id = ? AND password = ?", SInReq.UserId, SInReq.Password).First(&user).Error
+	if err != nil {
+		return SInRes, errors.New("sign in Failed")
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: Can we add explicitly about the error?

File: internal/services/user_management/user_management.go [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+	err = config.DB.Table("users").Where("user_id = ? AND email_id = ?", FPReq.UserId, FPReq.EmailId).First(&user).Error
+	if err != nil {
+		return FPRes, errors.New("user not found")
+	}
+	otp, err_ := redis.GetRandNum()
+	if err_ != nil {
+		return FPRes, errors.New("otp not generated")
+	}
+	FPRes.Otp = otp
+	e := redis.SetValue(FPReq.EmailId, otp, 5*time.Minute)
```
Comment (mukesh0513) [inline internal/services/user_management/user_management.go]: what is e? not good variable initialisation.

File: internal/services/user_management/user_management.go (hunk 11 above)
Comment (Ayushnit) [reply internal/services/user_management/user_management.go]: yeah in case no error comes, user registered message shows

**Comment by mukesh0513** [comment]
Check variable naming. Not very descriptive at some places to understand the use of variable.

File: internal/services/user_management/user_management.go (hunk 10 above)
Comment (Ayushnit) [reply internal/services/user_management/user_management.go]: ok

**Comment by mukesh0513** [comment]
Do not send successsful responses for error scenarios also.

File: internal/services/user_management/user_management.go (hunk 9 above)
Comment (Ayushnit) [reply internal/services/user_management/user_management.go]: i generate the user_id from the email_id

File: internal/provider/redis/redis.go (hunk 4 above)
Comment (Ayushnit) [reply internal/provider/redis/redis.go]: in the development.yaml file ?

File: internal/services/user_management/user_management.go (hunk 13 above)
Comment (Ayushnit) [reply internal/services/user_management/user_management.go]: ok, done

File: internal/provider/redis/redis.go [hunk 19]
```diff
@@ -0,0 +1,33 @@
+package redis
+
+import (
+	"Tradeasy/config"
+	"github.com/go-redis/redis"
+	"time"
+)
+
+func createClient() *redis.Client {
```
Comment (mukesh0513) [inline internal/provider/redis/redis.go]: why are we creating new client every time.

Comment (mukesh0513) [reply internal/provider/redis/redis.go]: check if redis connected succesfully or not.

**Review by mukesh0513 — APPROVED** [review]
LGTM
